apollo-router/src/query_planner/dual_query_planner.rs

```diff
@@ -318,25 +403,65 @@ fn vec_matches_as_set<T>(this: &[T], other: &[T], item_matches: impl Fn(&T, &T)
         })
 }
 
-fn plan_node_matches(this: &PlanNode, other: &PlanNode) -> bool {
+fn vec_matches_result_as_set<T>(
+    this: &[T],
+    other: &[T],
+    item_matches: impl Fn(&T, &T) -> bool,
+) -> Result<(), MatchFailure> {
+    // Set-inclusion test in both directions
+    check_match_eq!(this.len(), other.len());
+    for (index, this_node) in this.iter().enumerate() {
+        if !other
+            .iter()
+            .any(|other_node| item_matches(this_node, other_node))
+        {
+            return Err(MatchFailure::new(format!(
+                "mismatched set: missing item[{}]",
+                index
+            )));
+        }
+    }
+    for other_node in other.iter() {
+        if !this
+            .iter()
+            .any(|this_node| item_matches(this_node, other_node))
+        {
+            return Err(MatchFailure::new(
+                "mismatched set: extra item found".to_string(),
+            ));
+        }
+    }
+    assert!(vec_matches_as_set(this, other, item_matches));
+    Ok(())
+}
+
+fn plan_node_matches(this: &PlanNode, other: &PlanNode) -> Result<(), MatchFailure> {
     match (this, other) {
         (PlanNode::Sequence { nodes: this }, PlanNode::Sequence { nodes: other }) => {
-            vec_matches(this, other, plan_node_matches)
+            vec_matches_result(this, other, plan_node_matches)
+                .map_err(|err| err.add_description("under Sequence node"))?;
         }
         (PlanNode::Parallel { nodes: this }, PlanNode::Parallel { nodes: other }) => {
-            vec_matches_as_set(this, other, plan_node_matches)
+            vec_matches_result_as_set(this, other, |a, b| plan_node_matches(a, b).is_ok())
+                .map_err(|err| err.add_description("under Parallel node"))?;
+        }
+        (PlanNode::Fetch(this), PlanNode::Fetch(other)) => {
+            fetch_node_matches(this, other)
+                .map_err(|err| err.add_description("under Fetch node"))?;
```

Comment:
Might be nice to add the `this.operation_name` to the message

Reply:
good idea!

Reply:
Or would it make it harder to categorise differences? It's nice to have for an individual plan but maybe not for mass-comparisons with harness.

Reply:
I think I can filter out operation names in the classification script.